primitives/arithmetic/src/per_things.rs

```diff
@@ -1444,3 +1445,150 @@ implement_per_thing_with_perthousand!(
 	u128,
 	"_Parts per Quintillion_",
 );
+
+macro_rules! implement_checked_arithmetic {
+	(
+		$name:ident,
+		$test_mod:ident,
+		$type:ty,
+		$max:tt,
+	) => {
+		impl Add<Self> for $name {
+			type Output = $name;
+
+			#[inline]
+			fn add(self, rhs: Self) -> Self::Output {
+				$name::from_parts(self.deconstruct().add(rhs.deconstruct()))
```

Comment:
note that this from_parts will saturate in case it is more than max

Reply:
Do we want to panic here as opposed to saturation?

Reply:
probably

Reply:
Maybe we can do a debug assert, and keep saturating.